### ISSUE TITLE
Need a way to hook code into the ContentfulClient pipeline, via events or alternative methods

### ISSUE THREAD

**Issue by deanebarker** [contentful/contentful.net#4]
TLDR: we need events (-ish).

One of the problems with developing a "seamless" API is that there are, well, _no seams_.  And by "seams," I mean "places to inject code."

For example, I have some link resolution I need to do on my Term objects. In addition to processing the Markdown, I need to create some links, wiki-style.  It's a unique use case, but not _that_ unique, in the sense that it's not uncommon to need to do _something_ post-population with any object store.

When I was doing this with the last API, all I had were the fields and had to populate the POCO myself in a constructor. But since you do the object population for me now, I've "lost" that opportunity to hook the code. I get a fully-populated object, almost magically, in a sense before I'm "ready" for it.

What I'd like to see is either a full-blown event model, like this:

```
ContentfulClient.OnAfterObjectPopulated += ResolveLinks
```

The one drawback here is that this is global to the client, where I might need different handlers for different classes. (In my example use case: Terms need to resolve links; Chapters do not.)

Or even just an interface, like this:

```
// This is a terrible interface name, I know...
public interface IHasAfterPopulatedMethod
{
   void OnAfterPopulated();
}
```

If my POCO implemented that interface, you'd call `OnAfterPopulation` for me at the correct time.  The benefit of this is that this would be on a per-class basis.

Another, more wide-ranging, alternative might be a `ContentfulEntry` abstract class, from which my POCOs could inherit.  It would have a set of events on it, inherited by my POCOs, which I could hook individually:

```
Term.OnAfterObjectPopulated += ResolveLinks
```

(Now that I think about it, this could have the `SystemProperties` property on it from #3 as well. Looking down the road, I think there are multiple instances where a base class for the POCOs would come in handy. Putting the `SystemProperties` on it now might be enough of an inducement to get developers doing it with their early code.)

Perhaps do both: (1) have events on `ContentfulClient`, and (2) have a `ContentfulEntry` abstract base class with inherited events for individual POCOs.  This way, some code could be global, and other code would be per-class.
